jubilant/_juju.py

```diff
@@ -1297,6 +1324,43 @@ def _temp_dir(self) -> str:
         else:
             return tempfile.gettempdir()
 
+    # This context manager is for deploy() and refresh(), and automatically copies
+    # a local charm file and local resource files into a temporary directory if Juju
+    # is running as a snap (in which case /tmp is not accessible).
+    @contextlib.contextmanager
+    def _deploy_tempdir(
+        self,
+        charm: str | pathlib.Path | None,
+        resources: Mapping[str, str] | None,
+    ) -> Generator[tuple[str | None, Mapping[str, str] | None]]:
+        if charm is not None:
+            charm = str(charm)
+        charm_needs_temp = charm is not None and charm.startswith(('.', '/'))
```

Comment:
A path could also start with `~`, maybe other things as well? Perhaps we should [parse as a URL](https://github.com/juju/juju/blob/c19c662002d32b5710120084bd3bc2df4cf7c4b2/internal/charm/url.go#L139)? I'm also ok with starting simple and adjusting if charmers come to us with issues.

Reply:
I think it's the shell that expands `~`, so you wouldn't be expecting it in this context. Also, I believe `juju deploy` determines whether a charm is a local file by whether it starts with `.` or `/`, so I'm following that logic here. But you're right, if we need to adjust later we can.

Reply:
> I think it's the shell that expands `~`, so you wouldn't be expecting it in this context.

Ah, true.

> Also, I believe `juju deploy` determines whether a charm is a local file by whether it starts with `.` or `/`, so I'm following that logic here.

I could not find any Juju code that did that. I believe it's the code I linked above, which is parsing as a URL.